### PR TITLE
Migrate Streamable to Writeable for recovery package

### PR DESCRIPTION
**Pull request by liketic**
Migrate more classes to ```Writable```. 

Relates to #34389

### CODE REVIEW

**Comment by dnhatn** [comment]
@liketic Thanks for picking up this :). It would be easier to review if you migrate one class or one package per PR. Let narrow this PR to the "recovery" package only.

**Comment by liketic** [comment]
Thanks @dnhatn , 👍  I updated this as you requested. Could you please review again?

**Comment by dnhatn** [comment]
@liketic Thank you! I will have a look.
@elasticmachine ok to test.

**Comment by dnhatn** [comment]
@elasticmachine retest this please.

**Comment by dnhatn** [comment]
@elasticmachine retest this please.

File: server/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryRequest.java [hunk 1]
```diff
@@ -41,6 +41,12 @@ public RecoveryRequest() {
         this(Strings.EMPTY_ARRAY);
     }
 
+    public RecoveryRequest(StreamInput in) throws IOException {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryRequest.java]: Let's revert this change for having two variants is not a good idea.

File: server/src/main/java/org/elasticsearch/action/support/broadcast/BroadcastRequest.java [hunk 2]
```diff
@@ -36,6 +36,12 @@
     public BroadcastRequest() {
     }
 
+    public BroadcastRequest(StreamInput in) throws IOException {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/action/support/broadcast/BroadcastRequest.java]: Let's revert this change for having two variants is not a good idea.

File: server/src/main/java/org/elasticsearch/action/admin/indices/recovery/TransportRecoveryAction.java [hunk 3]
```diff
@@ -93,9 +93,7 @@ protected RecoveryResponse newResponse(RecoveryRequest request, int totalShards,
 
     @Override
     protected RecoveryRequest readRequestFrom(StreamInput in) throws IOException {
-        final RecoveryRequest recoveryRequest = new RecoveryRequest();
-        recoveryRequest.readFrom(in);
-        return recoveryRequest;
+        return new RecoveryRequest(in);
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/action/admin/indices/recovery/TransportRecoveryAction.java]: we also need to revert this change after reverting the RecoveryRequest class.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java [hunk 4]
```diff
@@ -347,12 +351,22 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         static final String TARGET_THROTTLE_TIME_IN_MILLIS = "target_throttle_time_in_millis";
     }
 
-    public static class Timer implements Streamable {
+    public static class Timer implements Streamable, Writeable {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java]: I think we can remove `Streamable` from `Timer` then remove `readFrom` entirely.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java [hunk 5]
```diff
@@ -414,9 +424,16 @@ public synchronized void writeTo(StreamOutput out) throws IOException {
 
     }
 
-    public static class VerifyIndex extends Timer implements ToXContentFragment, Streamable {
+    public static class VerifyIndex extends Timer implements ToXContentFragment, Streamable, Writeable {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java]: I think we can remove `Streamable` from `VerifyIndex` then remove `readFrom` entirely.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java [hunk 6]
```diff
@@ -451,13 +467,23 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         }
     }
 
-    public static class Translog extends Timer implements ToXContentFragment, Streamable {
+    public static class Translog extends Timer implements ToXContentFragment, Streamable, Writeable {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java]: I think we can remove `Streamable` from `Translog` then remove `readFrom` entirely.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java [hunk 7]
```diff
@@ -560,7 +583,7 @@ public synchronized XContentBuilder toXContent(XContentBuilder builder, Params p
         }
     }
 
-    public static class File implements ToXContentObject, Streamable {
+    public static class File implements ToXContentObject, Streamable, Writeable {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java]: I think we can remove `Streamable` from `File` then remove `readFrom` entirely.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java [hunk 8]
```diff
@@ -671,7 +692,7 @@ public String toString() {
         }
     }
 
-    public static class Index extends Timer implements ToXContentFragment, Streamable {
+    public static class Index extends Timer implements ToXContentFragment, Streamable, Writeable {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java]: I think we can remove `Streamable` from `Index` then remove `readFrom` entirely.

File: server/src/test/java/org/elasticsearch/indices/recovery/RecoveryTargetTests.java [hunk 9]
```diff
@@ -93,12 +93,10 @@ public T serializeDeserialize() throws IOException {
...
 
         protected T deserialize(StreamInput in) throws IOException {
-            T obj = createObj();
-            obj.readFrom(in);
-            return obj;
+            return createObj(in);
         }
 
-        abstract T createObj();
+        abstract T createObj(StreamInput in) throws IOException;
```
Comment (dnhatn) [inline server/src/test/java/org/elasticsearch/indices/recovery/RecoveryTargetTests.java]: We need to make T extends from Writable.

**Review by dnhatn — COMMENTED** [review]
Thanks again @liketic. I left some comments but this looks great. Would you please bring your PR up to date with the master branch?

File: server/src/test/java/org/elasticsearch/indices/recovery/RecoveryTargetTests.java (hunk 9 above)
Comment (liketic) [reply server/src/test/java/org/elasticsearch/indices/recovery/RecoveryTargetTests.java]: Thanks @dnhatn , I pushed 0bd175770a16c1bdd23b67186d0daadfd7accaef

**Comment by liketic** [comment]
@dnhatn I updated as you requested. Please review again.

**Review by dnhatn — APPROVED** [review]
@liketic I pushed https://github.com/elastic/elasticsearch/pull/37380/commits/f08269fac991c4a44821e73c3f729778dad1ff59 to arrange the ctors of RecoveryState. I'll wait for CI then pull this in. Thank you very much for your PR :).

**Comment by dnhatn** [comment]
I have merged this and backported to 6.x. Thanks again @liketic.
